File: internal/content/testdata/bench/csharp/nc-largest-rectangle.cs
using System.Collections.Generic;

public class Solution {
    public long largest_rectangle_area(long[] heights) {
        long maxArea = 0;
        var stack = new Stack<long[]>(); // [start_index, height]
        int n = heights.Length;
        for (int i = 0; i < n; i++) {
            long start = i;
            while (stack.Count > 0 && stack.Peek()[1] > heights[i]) {
                long[] top = stack.Pop();
                long idx = top[0], height = top[1];
                maxArea = Math.Max(maxArea, height * (i - idx));
                start = idx;
            }
            stack.Push(new long[] { start, heights[i] });
        }
        foreach (long[] entry in stack) {
            maxArea = Math.Max(maxArea, entry[1] * (n - entry[0]));
        }
        return maxArea;
    }
}

File: internal/content/testdata/bench/go/nc-largest-rectangle.go
func largest_rectangle_area(heights []int) int {
	maxArea := 0
	type pair struct{ start, height int }
	stack := []pair{}
	for i, h := range heights {
		start := i
		for len(stack) > 0 && stack[len(stack)-1].height > h {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			area := top.height * (i - top.start)
			if area > maxArea {
				maxArea = area
			}
			start = top.start
		}
		stack = append(stack, pair{start, h})
	}
	for _, p := range stack {
		area := p.height * (len(heights) - p.start)
		if area > maxArea {
			maxArea = area
		}
	}
	return maxArea
}
